[c10d] Enable CudaEventCache by default and add multi device support

fduwjj opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #140975

We added `CudaEventCache` in https://github.com/pytorch/pytorch/pull/133727 and this is a feature which tries to reuse CudaEvent so that we don't call destroy of CudaEvent which causes hang in the past. We had a bunch of tests and testing on TorchTitan and internal workload already. So far no errors or crash are found at the moment so we decide to roll out to all OSS users. For internal workload, this PR would not affect it because of some internal gating.

Also we observed some multi-device use cases in OSS, so that we want to bring back multi-device support originally proposed in https://github.com/pytorch/pytorch/pull/122732/files.

cc @H-Huang @awgu @kwen2501 @wanchaol @fegin @wz337 @wconstab @d4l3k @c-p-i-o

fduwjj committed:
[c10d] Enable CudaEventCache by default

[ghstack-poisoned]

eqy reviewed (commented):
Failures look real, perhaps there's an issue with the cache being shared across streams/devices?

fduwjj commented:
@eqy I cannot repro it locally...

kwen2501 reviewed (approved):
Approving. 
As discussed, a fix for the CI failure would be to create one cache per device index seen, as DDP still supports multi-device modules. (Though we may start deprecating it)

fduwjj committed:
Update on "[c10d] Enable CudaEventCache by default"


We added `CudaEventCache` in https://github.com/pytorch/pytorch/pull/133727 and this is a feature which tries to reuse CudaEvent so that we don't call destroy of CudaEvent which causes hang in the past. We had a bunch of tests and testing on TorchTitan and internal workload already. So far no errors or crash are found at the moment so we decide to roll out to all OSS users. For internal workload, this PR would not affect it because of some internal gating.

cc H-Huang awgu kwen2501 wanchaol fegin wz337 wconstab d4l3k c-p-i-o

[ghstack-poisoned]

fduwjj committed:
Update on "[c10d] Enable CudaEventCache by default"


We added `CudaEventCache` in https://github.com/pytorch/pytorch/pull/133727 and this is a feature which tries to reuse CudaEvent so that we don't call destroy of CudaEvent which causes hang in the past. We had a bunch of tests and testing on TorchTitan and internal workload already. So far no errors or crash are found at the moment so we decide to roll out to all OSS users. For internal workload, this PR would not affect it because of some internal gating.

cc H-Huang awgu kwen2501 wanchaol fegin wz337 wconstab d4l3k c-p-i-o

[ghstack-poisoned]

kwen2501 commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -796,25 +797,29 @@ void ProcessGroupNCCL::WorkNCCL::abort() {
 }
 
-ProcessGroupNCCL::CUDAEventCache::CUDAEventCache() = default;
+ProcessGroupNCCL::CUDAEventCache::CUDAEventCache()
+    : caches_(at::cuda::device_count()) {}
```
If we use map, we wouldn't need to allocate many up front?

fduwjj replied:
Well, when you have many devices, you are going to allocate anyway.. 

Also computing hash is more expensive than index look-up, lol

kwen2501 replied:
I disagree, here we are allocating `device_count()` many, whether they are used or not. 
Whereas, in most cases, we only use 1, maybe in corner cases 2.

fduwjj replied:
What is the value of `device_count()`? Isn't it always 1?

kwen2501 replied:
And after we retire DDP's multi-device support, we will have to rework this code if it is as-is.

fduwjj committed:
Update on "[c10d] Enable CudaEventCache by default and add multi device support"


We added `CudaEventCache` in https://github.com/pytorch/pytorch/pull/133727 and this is a feature which tries to reuse CudaEvent so that we don't call destroy of CudaEvent which causes hang in the past. We had a bunch of tests and testing on TorchTitan and internal workload already. So far no errors or crash are found at the moment so we decide to roll out to all OSS users. For internal workload, this PR would not affect it because of some internal gating.

Also we observed some multi-device use cases in OSS, so that we want to bring back multi-device support originally proposed in https://github.com/pytorch/pytorch/pull/122732/files.

cc H-Huang awgu kwen2501 wanchaol fegin wz337 wconstab d4l3k c-p-i-o

[ghstack-poisoned]

fduwjj commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -798,6 +799,12 @@ void ProcessGroupNCCL::WorkNCCL::abort() {
 
 ProcessGroupNCCL::CUDAEventCache::CUDAEventCache() = default;
 
+ProcessGroupNCCL::CUDAEventCache::CUDAEventCache(
+    CUDAEventCache&& cache) noexcept = default;
+
+ProcessGroupNCCL::CUDAEventCache& ProcessGroupNCCL::CUDAEventCache::
+    CUDAEventCache::operator=(CUDAEventCache&& cache) noexcept = default;
+
```
I am wondering if we want to move L800-L807 to hpp?

fduwjj commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.hpp:
```diff
@@ -459,11 +459,13 @@ class TORCH_API ProcessGroupNCCL : public Backend {
 
    private:
-    std::mutex cacheMutex_;
+    std::unique_ptr<std::mutex> cacheMutexPtr_;
```
std::mutex is not copyable and movable...

fduwjj committed:
Update on "[c10d] Enable CudaEventCache by default and add multi device support"


We added `CudaEventCache` in https://github.com/pytorch/pytorch/pull/133727 and this is a feature which tries to reuse CudaEvent so that we don't call destroy of CudaEvent which causes hang in the past. We had a bunch of tests and testing on TorchTitan and internal workload already. So far no errors or crash are found at the moment so we decide to roll out to all OSS users. For internal workload, this PR would not affect it because of some internal gating.

Also we observed some multi-device use cases in OSS, so that we want to bring back multi-device support originally proposed in https://github.com/pytorch/pytorch/pull/122732/files.

cc H-Huang awgu kwen2501 wanchaol fegin wz337 wconstab d4l3k c-p-i-o

[ghstack-poisoned]

fduwjj committed:
Update on "[c10d] Enable CudaEventCache by default and add multi device support"


We added `CudaEventCache` in https://github.com/pytorch/pytorch/pull/133727 and this is a feature which tries to reuse CudaEvent so that we don't call destroy of CudaEvent which causes hang in the past. We had a bunch of tests and testing on TorchTitan and internal workload already. So far no errors or crash are found at the moment so we decide to roll out to all OSS users. For internal workload, this PR would not affect it because of some internal gating.

Also we observed some multi-device use cases in OSS, so that we want to bring back multi-device support originally proposed in https://github.com/pytorch/pytorch/pull/122732/files.

cc H-Huang awgu kwen2501 wanchaol fegin wz337 wconstab d4l3k c-p-i-o

[ghstack-poisoned]

fduwjj committed:
Update on "[c10d] Enable CudaEventCache by default and add multi device support"


We added `CudaEventCache` in https://github.com/pytorch/pytorch/pull/133727 and this is a feature which tries to reuse CudaEvent so that we don't call destroy of CudaEvent which causes hang in the past. We had a bunch of tests and testing on TorchTitan and internal workload already. So far no errors or crash are found at the moment so we decide to roll out to all OSS users. For internal workload, this PR would not affect it because of some internal gating.

Also we observed some multi-device use cases in OSS, so that we want to bring back multi-device support originally proposed in https://github.com/pytorch/pytorch/pull/122732/files.

cc H-Huang awgu kwen2501 wanchaol fegin wz337 wconstab d4l3k c-p-i-o

[ghstack-poisoned]

fduwjj committed:
Update

[ghstack-poisoned]

fduwjj commented:
@pytorchbot merge
